src/components/CredentialCard/CredentialCard.tsx

```diff
@@ -110,6 +118,242 @@ function CredentialName({
+        )}
+    </Fragment>
+  );
+}
+
+function DownloadModal({
+  credential,
+  portalRef,
+}: {
+  credential: Credential;
+  portalRef: RefObject<HTMLElement>;
+}) {
+  const t = browser.i18n.getMessage;
+  const visibility = useBooleanState();
+
+  const markCredentialDownloaded = useCallback(async () => {
+    await saveCredential({ ...credential, isDownloaded: true });
+  }, [credential]);
+
+  const [checked, setChecked] = useState(false);
+
+  const handleToggle = useCallback((event) => {
+    setChecked(event.target.checked);
+  }, []);
+
+  const handleConfirm = useCallback(async () => {
+    await markCredentialDownloaded();
+    await setShowDownloadInfo(!checked);
+    visibility.off();
+  }, [markCredentialDownloaded, checked, visibility]);
+
+  const showInfo = useShowDownloadInfo();
+  const { isDownloaded } = credential;
+  const { name, url } = getCredentialDownload(credential);
+
+  return (
+    <Fragment>
+      {!isDownloaded && showInfo ? (
+        <button
+          className={
+            isDownloaded ? styles.download : styles.downloadPromptExpanded
+          }
+          aria-label={t('component_CredentialCard_backup')}
+          onClick={visibility.on}
+        />
+      ) : (
+        <a
+          download={name}
+          href={url}
+          aria-label={t('component_CredentialCard_backup')}
+          className={
+            isDownloaded ? styles.download : styles.downloadPromptExpanded
+          }
+          onClick={markCredentialDownloaded}
+        />
+      )}
+
+      {visibility.current &&
+        portalRef.current &&
+        createPortal(
+          <Modal open className={styles.overlay}>
+            <h2 className={styles.downloadInfo}>
+              {t('component_CredentialCard_download_info')}
+            </h2>
+            <a
+              download={name}
+              href={url}
+              className={styles.confirmDownload}
+              onClick={handleConfirm}
+            >
+              {t('component_CredentialCard_download_confirm')}
+            </a>
+            <button
+              type="button"
+              className={styles.cancelDownload}
+              onClick={visibility.off}
+            >
+              {t('common_action_close')}
+            </button>
+            <label className={styles.toggleLabel}>
+              {t('component_CredentialCard_download_toggle')}
+              <input
+                className={styles.toggle}
+                type="checkbox"
+                defaultChecked={false}
+                onClick={handleToggle}
+              />
+              <span />
+            </label>
+          </Modal>,
+          portalRef.current,
+        )}
+    </Fragment>
+  );
+}
+
+function PresentationModal({
+  credential,
+  portalRef,
+}: {
+  credential: Credential;
+  portalRef: RefObject<HTMLElement>;
+}) {
+  const t = browser.i18n.getMessage;
+  const visibility = useBooleanState();
+
+  const [checked, setChecked] = useState(false);
```

Comment:
Can we use `useBooleanState` here too?

Reply:
I considered that, but here it won’t even save us a single line of code, right?